### PR TITLE
Issue #13809: Kill mutation in XpathUtil

### PR DESCRIPTION
**Pull request by Kevin222004**
Issue #13809: Kill mutation in XpathUtil

https://github.com/checkstyle/checkstyle/blob/0322ba1be8489d1df373013efc9078de4d209eb4/config/pitest-suppressions/pitest-utils-suppressions.xml#L354-L361

### CODE REVIEW

**Review by nrmancuso — CHANGES_REQUESTED** [review]
We need to understand how pitest is actually mutating this code before we make changes like this. The code proposed by this PR is objectively less clear and concise than what we have now. We should not allow pitest to decrease our code quality, especially if we don’t even understand what it is doing under the hood.

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/XpathUtil.java [hunk 1]
```diff
@@ -214,8 +214,11 @@ public static List<NodeInfo> getXpathItems(String xpath, AbstractNode rootNode)
         final XPathDynamicContext xpathDynamicContext = xpathExpression
                 .createDynamicContext(rootNode);
         final List<Item> items = xpathExpression.evaluate(xpathDynamicContext);
-        return items.stream()
-                .map(NodeInfo.class::cast)
-                .collect(Collectors.toUnmodifiableList());
+        final List<NodeInfo> nodeInfoList = new ArrayList<>();
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/XpathUtil.java:217]: please make method `UnmodifiableCollectionUtil.unmodifiableList(items, NodeInfo.class)`
to keep what pitest naturaly see as not required code is special place/class.

**Review by romani — CHANGES_REQUESTED** [review]
items:

**Review by romani — APPROVED** [review]
Ok to merge.

@Vyom-Yadav , please share your opinion on cast problem, looks like this becomes hot topic.

**Comment by romani** [comment]
@Kevin222004 , please rebase to resolve conflict

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java [hunk 2]
```diff
@@ -63,6 +67,20 @@ public static <T> List<T> unmodifiableList(List<T> collection) {
...
 
+    /**
+     * Returns an unmodifiable view of a List containing elements of a specific type.
+     *
+     * @param items The List of items to make unmodifiable.
+     * @param elementType The Class object representing the type of elements in the list.
+     * @param <T> The type of elements in the resulting unmodifiable List.
+     * @return An unmodifiable List containing elements of the specified type.
+     */
+    public static <T> List<T> unmodifiableList(Collection<Item> items, Class<T> elementType) {
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java:78]: Why do you use exact type here (which is `Item`). This method is supposed to be generic. Can we update it as follows?
```
public static <S, T> List<T> unmodifiableList(Collection<S> items, Class<T> elementType) { ... }
```
here, `S` means source type and `T` means target type.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java:78]: @Kevin222004 , please fix as suggested.

Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java]: done

Comment (rdiachenko) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java]: Do we need this block after the above change? Could you remove it if not needed?
```
    <file name="UnmodifiableCollectionUtil">
       <allow pkg="net.sf.saxon"/>
     </file>
```

Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java]: @rdiachenko removed.

**Comment by romani** [comment]
@Kevin222004 , please address Checker

**Review by rdiachenko — APPROVED** [review]
lgtm

@Kevin222004 please resolve checker errors

**Review by Vyom-Yadav — APPROVED** [review]
LGTM! Please fix CI errors.

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java [hunk 3]
```diff
@@ -63,6 +65,21 @@ public static <T> List<T> unmodifiableList(List<T> collection) {
...
+     * @param elementType The Class object representing the type of elements in the list.
+     * @param <S> The generic type of elements in the input Collection.
+     * @param <T> The type of elements in the resulting unmodifiable List.
+     * @return An unmodifiable List containing elements of the specified type.
+     */
+    public static <S, T> List<T> unmodifiableList(Collection<S> items, Class<T> elementType) {
+        return items.stream()
+                .map(elementType::cast)
+                .collect(Collectors.toUnmodifiableList());
+    }
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java:81]: Why did we make this about collections, when the issue is with the calls to `Class#cast`?  If we want to make some hack to bypass pitest, why not just have a util method to cast the stream as an intermediate operation?

Pseudocode:
```
<R, T> Stream<T> castedStreamOF(Stream<R> s, Class<T> c) {
  return stream.map(c::cast)
}

```

Then we can collect or perform subsequent stream operations elsewhere more readily.

**Review by nrmancuso — DISMISSED** [review]
Item:

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java (hunk 3 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java:81]: It has unmodified collection, that pitest will complain about, not sure why it is not doing it now.

And current implementation is easier, and we don't want to make it highly reusable, at least for now.

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java:81]: Why is it a public method in a utility class if we don't want it to be reusable?

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java:81]: Pitest limitation to allow easy exclude from mutation by class name

**Review by nrmancuso — COMMENTED** [review]
The changes here don't add up for me. In order to remove one pitest suppression (that should just be permanent IMO), we have added:
 - 2 checker suppressions (tech debt)
 - avoidCallsTo suppression in pom.xml (tech debt)
 - yet another specialized utility method that won't get used anywhere else (tech debt)
 - Production code (the reason why we are really here) that is much less clear than before

At this point, we are shuffling tech debt around. I won't block this, but I cannot approve either.

**Comment by nrmancuso** [comment]
@romani let’s just merge this and move on. We have already spent too much time and resources on this PR.

**Pull request merged**